File: src/main/java/minic/ui/text/MiniCLineTokenHighlighter.java
package minic.ui.text;

import java.util.ArrayList;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

final class MiniCLineTokenHighlighter {
    private static final Pattern TOKEN = Pattern.compile(
            "[-+]?0x[0-9A-Fa-f]+|[-+]?[0-9]+(?:\\.[0-9]+)?|[%@$&.]?[A-Za-z0-9_][A-Za-z0-9_.$]*|\\S"
    );

    private MiniCLineTokenHighlighter() {}

    static List<MiniCStyledTextSegment> highlight(String line, TokenClassifier classifier) {
        String text = line == null || line.isEmpty() ? " " : line;
        ArrayList<MiniCStyledTextSegment> segments = new ArrayList<>();
        Matcher matcher = TOKEN.matcher(text);
        int cursor = 0;
        while (matcher.find()) {
            if (matcher.start() > cursor) {
                add(segments, text.substring(cursor, matcher.start()), MiniCTextStyleRole.CODE_PLAIN);
            }
            String token = matcher.group();
            add(segments, token, classifier.roleFor(token, matcher.start(), text));
            cursor = matcher.end();
        }
        if (cursor < text.length()) {
            add(segments, text.substring(cursor), MiniCTextStyleRole.CODE_PLAIN);
        }
        return List.copyOf(segments);
    }

    static void add(List<MiniCStyledTextSegment> segments, String text, MiniCTextStyleRole role) {
        if (text.isEmpty()) {
            return;
        }
        if (!segments.isEmpty()) {
            MiniCStyledTextSegment previous = segments.getLast();
            if (previous.role() == role) {
                segments.set(segments.size() - 1, new MiniCStyledTextSegment(previous.text() + text, role));
                return;
            }
        }
        segments.add(new MiniCStyledTextSegment(text, role));
    }

    interface TokenClassifier {
        MiniCTextStyleRole roleFor(String token, int startOffset, String fullLine);
    }
}

File: uiweb/src/text/MiniCLineTokenHighlighter.ts
import type { JavaMirrorFile } from "../translation/javaMirror";
import { MiniCStyledTextSegment } from "./MiniCStyledTextSegment";
import { MiniCTextStyleRole } from "./MiniCTextStyleRole";
import type { MiniCTextStyleRole as MiniCTextStyleRoleValue } from "./MiniCTextStyleRole";

export const miniCLineTokenHighlighterMirror = {
  javaPath: "src/main/java/minic/uilocal/text/MiniCLineTokenHighlighter.java",
  webPath: "uiweb/src/text/MiniCLineTokenHighlighter.ts",
  packageName: "minic.uilocal.text",
  exportName: "MiniCLineTokenHighlighter",
  kind: "class",
  imports: ["java.util.ArrayList", "java.util.List", "java.util.regex.Pattern"],
  fields: [{ name: "TOKEN", signature: "private static final Pattern TOKEN =" }],
  methods: [
    { name: "highlight", signature: "highlight(String line, TokenClassifier classifier)" },
    { name: "add", signature: "add(List<MiniCStyledTextSegment> segments, String text, MiniCTextStyleRole role)" },
  ],
} as const satisfies JavaMirrorFile;

export interface MiniCTokenClassifier {
  roleFor(token: string, startOffset: number, fullLine: string): MiniCTextStyleRoleValue;
}

const TOKEN_PATTERN = /[-+]?0x[0-9A-Fa-f]+|[-+]?[0-9]+(?:\.[0-9]+)?|[%@$&.]?[A-Za-z0-9_][A-Za-z0-9_.$]*|\S/g;

export class MiniCLineTokenHighlighter {
  static readonly mirror = miniCLineTokenHighlighterMirror;

  static highlight(line: string | null | undefined, classifier: MiniCTokenClassifier): readonly MiniCStyledTextSegment[] {
    const text = line && line.length > 0 ? line : " ";
    const segments: MiniCStyledTextSegment[] = [];
    let cursor = 0;

    for (const match of text.matchAll(TOKEN_PATTERN)) {
      const token = match[0];
      const start = match.index ?? 0;
      if (start > cursor) {
        this.add(segments, text.slice(cursor, start), MiniCTextStyleRole.CODE_PLAIN);
      }
      this.add(segments, token, classifier.roleFor(token, start, text));
      cursor = start + token.length;
    }

    if (cursor < text.length) {
      this.add(segments, text.slice(cursor), MiniCTextStyleRole.CODE_PLAIN);
    }

    return segments;
  }

  static add(
    segments: MiniCStyledTextSegment[],
    text: string,
    role: MiniCTextStyleRoleValue,
  ): void {
    if (text.length === 0) {
      return;
    }
    const previous = segments.at(-1);
    if (previous && previous.role === role) {
      segments[segments.length - 1] = new MiniCStyledTextSegment(previous.text + text, role);
      return;
    }
    segments.push(new MiniCStyledTextSegment(text, role));
  }
}

export default MiniCLineTokenHighlighter;
